### PR TITLE
PRMP - 841 bulk report

### PR DESCRIPTION
**Pull request by steph-torres-nhs**

### CODE REVIEW

File: lambdas/services/bulk_upload_report_service.py [hunk 1]
```diff
@@ -67,18 +221,15 @@ def get_dynamodb_report_items(
...
-                dict_writer_object.writerow(item)
-
     @staticmethod
     def get_times_for_scan() -> tuple[datetime, datetime]:
+        # current_time = datetime.datetime.now()
+        # today_date = datetime.datetime.today()
+        # start_timestamp = today_date - datetime.timedelta(days=30)
+        # start_timestamp = datetime.datetime.combine(start_timestamp, datetime.time.min)
+        # end_timestamp = current_time
+        # return start_timestamp, end_timestamp
```
Comment (steph-torres-nhs) [inline lambdas/services/bulk_upload_report_service.py:231]: left in to help make testing easier

File: lambdas/services/bulk_upload_report_service.py [hunk 2]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+        failures_per_patient = {}
+        unique_failures = {}
+
+        logger.info(f"Generating ODS report file for {ods_code}")
+
+        for item in ods_report_data:
+            nhs_number = item.get(MetadataReport.NhsNumber)
+            upload_status = item.get(MetadataReport.UploadStatus)
+            pds_ods_code = item.get(MetadataReport.PdsOdsCode)
+            uploader_ods_code = item.get(MetadataReport.UploaderOdsCode)
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:77]: Can we use 'BulkUploadReport' model here?

File: lambdas/models/bulk_upload_status.py [hunk 3]
```diff
@@ -20,18 +20,35 @@ class BulkUploadReport(BaseModel):
...
+    def __eq__(self, other):
+        if isinstance(other, OdsReport):
+            return (
+                self.ods_code == other.ods_code
+                and self.total_successful == other.total_successful
+                and self.total_registered_elsewhere == other.total_registered_elsewhere
+                and self.total_suspended == other.total_suspended
+                and self.failure_reasons == other.failure_reasons
+            )
+        return False
```
Comment (NogaNHS) [inline lambdas/models/bulk_upload_status.py:54]: Can we use pydantic here?

File: lambdas/services/bulk_upload_report_service.py [hunk 4]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+        for item in report_data:
+            uploader_ods_code = item.get(MetadataReport.UploaderOdsCode)
 
-            self.s3_service.upload_file(
-                s3_bucket_name=staging_bucket_name,
-                file_key=f"reports/{file_name}",
-                file_name=f"/tmp/{file_name}",
+            if uploader_ods_code not in grouped_ods_data:
+                grouped_ods_data[uploader_ods_code] = []
+            grouped_ods_data[uploader_ods_code].append(item)
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:52]: grouped_ods_data.setdefault(uploader_ods_code, []).append(item)

File: lambdas/services/bulk_upload_report_service.py [hunk 5]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+
+        logger.info(f"Generating ODS report file for {ods_code}")
+
+        for item in ods_report_data:
+            nhs_number = item.get(MetadataReport.NhsNumber)
+            upload_status = item.get(MetadataReport.UploadStatus)
+            pds_ods_code = item.get(MetadataReport.PdsOdsCode)
+            uploader_ods_code = item.get(MetadataReport.UploaderOdsCode)
+
+            if upload_status == "complete":
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:79]: UploadStatus enum class

File: lambdas/services/bulk_upload_report_service.py [hunk 6]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+                total_successful.add(nhs_number)
+
+                if uploader_ods_code != pds_ods_code:
+                    total_registered_elsewhere.add(nhs_number)
+
+            if upload_status == "suspended":
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:85]: upload_status == "suspended"? is that a upload status?

File: lambdas/services/bulk_upload_report_service.py [hunk 7]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+
+    def generate_ods_reports(
+        self, report_data: list[dict], generated_at: str
+    ) -> list[OdsReport]:
+        ods_reports: list[OdsReport] = []
 
-            logger.info("Uploading new report file to S3")
+        grouped_ods_data = {}
+        for item in report_data:
+            uploader_ods_code = item.get(MetadataReport.UploaderOdsCode)
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:48]: What if there no Uploader ODS code?

File: lambdas/services/bulk_upload_report_service.py [hunk 8]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+
+                if uploader_ods_code != pds_ods_code:
+                    total_registered_elsewhere.add(nhs_number)
+
+            if upload_status == "suspended":
+                total_suspended.add(nhs_number)
+
+            failure_reason = item.get(MetadataReport.FailureReason, "")
+            if failure_reason:
+                if nhs_number not in unique_failures:
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:90]:
```suggestion
            if failure_reason and nhs_number not in unique_failures:
```

Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:90]: Change unique_failures to a more meaningful name

File: lambdas/services/bulk_upload_report_service.py [hunk 9]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+                    unique_failures.update({nhs_number: failure_reason})
+
+        for reason in unique_failures.values():
+            if reason not in failures_per_patient:
+                failures_per_patient[reason] = 0
+            failures_per_patient[reason] += 1
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:96]: `for reason in unique_failures.values():
    failures_per_patient[reason] = failures_per_patient.get(reason, 0) + 1`

Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:96]:
```suggestion
for reason in unique_failures.values():
    failures_per_patient[reason] = failures_per_patient.get(reason, 0) + 1
```

File: lambdas/services/bulk_upload_report_service.py [hunk 10]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+
+        ods_report = OdsReport(
+            ods_code=ods_code,
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:99]: Change name to uploader ods

File: lambdas/services/bulk_upload_report_service.py [hunk 11]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
-                s3_bucket_name=staging_bucket_name,
-                file_key=f"reports/{file_name}",
-                file_name=f"/tmp/{file_name}",
+            if uploader_ods_code not in grouped_ods_data:
+                grouped_ods_data[uploader_ods_code] = []
+            grouped_ods_data[uploader_ods_code].append(item)
+
+        for ods_code, ods_data in grouped_ods_data.items():
+            ods_report = self.generate_individual_ods_report(
+                ods_code, grouped_ods_data[ods_code], generated_at
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:56]:
```suggestion
            ods_report = self.generate_individual_ods_report(
                ods_code, grouped_ods_data[ods_code], generated_at
```
```suggestion
        for ods_code, ods_data in grouped_ods_data.items():
            ods_report = self.generate_individual_ods_report(
                ods_code, ods_data, generated_at
```

File: lambdas/services/bulk_upload_report_service.py [hunk 12]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
 
-            self.s3_service.upload_file(
-                s3_bucket_name=staging_bucket_name,
-                file_key=f"reports/{file_name}",
-                file_name=f"/tmp/{file_name}",
+            if uploader_ods_code not in grouped_ods_data:
+                grouped_ods_data[uploader_ods_code] = []
+            grouped_ods_data[uploader_ods_code].append(item)
+
+        for ods_code, ods_data in grouped_ods_data.items():
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:54]: ods_code to uploader_ods_code

File: lambdas/services/bulk_upload_report_service.py [hunk 13]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+                writer.writerow(row)
+
+    def group_by_uploader_ods_code(self, report_data):
+        ods_reports = {}
+        for item in report_data:
+            ods_code = item.get(MetadataReport.UploaderOdsCode, "Unknown")
+            if ods_code not in ods_reports:
+                ods_reports[ods_code] = []
+            ods_reports[ods_code].append(item)
+        return ods_reports
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:201]: You don't use this fun

File: lambdas/tests/unit/services/test_bulk_upload_report_service.py [hunk 14]
```diff
@@ -131,91 +194,172 @@ def test_get_dynamo_data_with_no_start_key(mocker, set_env, bulk_upload_report_s
...
+        {"File name not matching Lloyd George naming convention": 1},
     )
-    mock_write_csv = mocker.patch(
-        "services.bulk_upload_report_service.BulkUploadReportService.write_items_to_csv"
+    actual = bulk_upload_report_service.generate_individual_ods_report(
+        TEST_CURRENT_GP_ODS, mock_ods_report_data, MOCK_END_REPORT_TIME
     )
 
-    bulk_upload_report_service.report_handler()
+    assert actual.__eq__(expected)
```
Comment (NogaNHS) [inline lambdas/tests/unit/services/test_bulk_upload_report_service.py:290]: obj1.__dict__ == obj2.__dict__

File: lambdas/models/bulk_upload_status.py [hunk 15]
```diff
@@ -20,18 +20,35 @@ class BulkUploadReport(BaseModel):
...
+        total_suspended=0,
+        failure_reasons=None,
+    ):
+        self.ods_code = ods_code
+        self.total_successful = total_successful
+        self.total_registered_elsewhere = total_registered_elsewhere
+        self.total_suspended = total_suspended
+
+        if failure_reasons is None:
+            failure_reasons = {}
```
Comment (NogaNHS) [inline lambdas/models/bulk_upload_status.py:42]: Set dict as default instead of None

File: lambdas/models/bulk_upload_status.py [hunk 16]
```diff
@@ -20,18 +20,35 @@ class BulkUploadReport(BaseModel):
...
+
+
+class OdsReport:
+    def __init__(
+        self,
+        ods_code: str,
+        total_successful=0,
+        total_registered_elsewhere=0,
+        total_suspended=0,
+        failure_reasons=None,
```
Comment (NogaNHS) [inline lambdas/models/bulk_upload_status.py:34]: Add types

File: lambdas/services/bulk_upload_report_service.py [hunk 17]
```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
...
+                total_suspended.add(nhs_number)
+
+            failure_reason = item.get(MetadataReport.FailureReason, "")
+            if failure_reason:
+                if nhs_number not in unique_failures:
+                    unique_failures.update({nhs_number: failure_reason})
+
+        for reason in unique_failures.values():
+            if reason not in failures_per_patient:
+                failures_per_patient[reason] = 0
```
Comment (NogaNHS) [inline lambdas/services/bulk_upload_report_service.py:95]: failures_per_patient[reason] = 1

**Review by danielleandersonNHS — APPROVED** [review]
Approved, any further changes to be picked up in new ticket

**Pull request merged**